neurobooth_os/iout/iphone.py

```diff
@@ -160,318 +130,430 @@ class IPhone:
-            bytes_to_pull = n
-            if (n - buff_recv) < MAX_RECV:
-                bytes_to_pull = n - buff_recv
-            packet = sock.recv(bytes_to_pull)
-            # packet = self._socket_recv(bytes_to_pull)
+            bytes_to_pull = n - bytes_received
+            if bytes_to_pull > MAX_RECV:
+                bytes_to_pull = MAX_RECV
 
-            buff_recv += len(packet)
+            packet = sock.recv(bytes_to_pull)
+            bytes_received += len(packet)
             fragments.append(packet)
-            if buff_recv >= n:
+
+            if bytes_received >= n:
                 break
-        data = b"".join(fragments)
-        return data
-
-    def _getpacket(self, timeout_in_seconds=20):
-        ready, _, _ = select.select([self.sock], [], [], timeout_in_seconds)
-        # print(ready)
-        if ready:
-            # first_frame=self._socket_recv(16)
-            first_frame = self.sock.recv(16)
-            version, type, tag, payload_size = struct.unpack("!IIII", first_frame)
-
-            if tag == 1 or tag == 2:
-                self._validate_message({}, tag)
-                payload = self.recvall(self.sock, payload_size)
-                return payload, version, type, tag
-            else:
-                # payload= self._socket_recv(payload_size)
-                payload = self.sock.recv(payload_size)
 
-            msg = self._json_unwrap(payload)
-            self._validate_message(msg, tag)
-            return msg, version, type, tag
+        return b"".join(fragments)
+
+    def _get_packet(self, timeout_sec: int = 20) -> PACKET_CONTENTS:
+        """
+        Retrieve a packet of data from the iPhone. This method will block until timed out.
+
+        :param timeout_sec: How long to block before timing out
+        :returns: (payload, version, type, tag): Payload is either a message dictionary (tag == 0) or byte string
+            (tag == 1 or tag == 2).
+        """
+        ready, _, _ = select.select([self.sock], [], [], timeout_sec)
+        if not ready:
+            self.logger.error(f"iPhone [state={self._state}]: Exceeded timeout for packet receipt")
+            raise IPhoneError(f"Timeout for packet receive exceeded ({timeout_sec} sec)")
+
+        first_frame = self.sock.recv(16)
+        version, type_, tag, payload_size = struct.unpack("!IIII", first_frame)
+
+        if tag in (1, 2):
+            payload = IPhone.recvall(self.sock, payload_size)
+            return payload, version, type_, tag
+        elif tag == 0:
+            payload = self.sock.recv(payload_size)
+            msg = IPhone._json_unwrap(payload)
+            self._validate_message(msg)
+            return msg, version, type_, tag
         else:
-            self.logger.error(f"iPhone: Exceed timeout for packet receipt")
-            raise IPhoneError(
-                f"Timeout for packet receive exceeded ({timeout_in_seconds} sec)"
-            )
+            self.logger.error('iPhone: PANIC')
```

Comment:
The idea with the IPhonePanic is that it is eventually handled by the panic() method, which log the exception message. So we should see a 'iPhone: PANIC' entry in the log followed by the f'iPhone [state={self._state}]: PANIC Message: {e}' entry. The initial message is so that the log captures the original function and line number of the event. I set this up this way so that we have mostly generic handling of these errors.

Reply:
Before, there was basically the same handling duplicated at every place with immediate returns and much more handling of success flags